### PR TITLE
 [addons][game] change addon system to C++ style

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description
This changes the Game addon system to the C ++ type.

In contrast to the other types of addon converted, this CPP instance type can only be used individually and not with others (at least for the moment), but this does not seem to be bad.

On Kodi's cpp page, the changes are also relatively small, only the addon header includes a little more.

## Motivation and Context
To finally have the old addon style out there more order in the headers.

## TODO
Add documentation about addon interface

The instance path is not (yet) on game system. Do not know me well enough to change that. Therefore, combination with other instance types is not yet possible. 
**But not necessary here and can be done separate.**

## How Has This Been Tested?
With game addons under Linux.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by Rechi** [comment]
Please run `git-clang-format` or `clang-format-diff` for every commit. It is annoying that I always have to comment about code style issues.

**Comment by AlwinEsch** [comment]
> Please run `git-clang-format` or `clang-format-diff` for every commit. It is annoying that I always have to comment about code style issues.

Is done

**Comment by AlwinEsch** [comment]
Have in the last commit the Doxygen documentation introduced.

see: http://alwinesch.github.io/group__cpp__kodi__addon__game.html

There are still some "/// @brief" empty for the structures in the header. @garbear habe you idea which text can be added there?

**Comment by garbear** [comment]
sorry I missed this! I'll make time for review this week

**Review by garbear — APPROVED** [review]

**Comment by garbear** [comment]
looks good! just leave the blank fields as is. the next time I update the libretro API I'll make sure everything is filled in.

**Comment by eigendude** [comment]
runtime tesed on macos with https://github.com/kodi-game/game.libretro/pull/54. jenkins failed in macOS test case, let's try again

jenkins build this please

**Comment by garbear** [comment]
ios64 failed after 11 seconds. jenkins build this please

**Comment by garbear** [comment]
tvos failed after 11 seconds. jenkins build this please

**Comment by garbear** [comment]
all succeeded but the expected uwp python3 error! can't wait for CI to go green so we can merge this. thanks @AlwinEsch and @Rechi!

File: xbmc/games/addons/GameClientInGameSaves.cpp [hunk 1]
```diff
@@ -26,9 +26,10 @@ using namespace GAME;
 #define INGAME_SAVES_EXTENSION_SAVE_RAM ".sav"
 #define INGAME_SAVES_EXTENSION_RTC      ".rtc"
 
-CGameClientInGameSaves::CGameClientInGameSaves(CGameClient* addon, const KodiToAddonFuncTable_Game* dllStruct) :
-  m_gameClient(addon),
-  m_dllStruct(dllStruct)
+CGameClientInGameSaves::CGameClientInGameSaves(CGameClient* addon,
+                                               const AddonInstance_Game* dllStruct)
+  : m_gameClient(addon)
+  , m_dllStruct(dllStruct)
```
Comment (Rechi) [inline xbmc/games/addons/GameClientInGameSaves.cpp]: > For lines up to [line length](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#line-length) everything stays on one line, excluding the braces which must be on the following lines.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#116-constructor-initialzation-lists

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Game.h [hunk 2]
```diff
@@ -0,0 +1,2360 @@
...
+  ///
+  virtual GAME_ERROR Serialize(uint8_t* data, size_t size)
+  {
+    return GAME_ERROR_NOT_IMPLEMENTED;
+  }
+  //----------------------------------------------------------------------------
+
+  //============================================================================
+  ///
+  /// @brief Deserialize the game from the given state
```
Comment (Rechi) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Game.h]: > The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Comment (AlwinEsch) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Game.h]: Do not quite understand, Sorry. You mean to increase the length if `===` to 100 there?

File: xbmc/games/addons/GameClientInGameSaves.cpp (hunk 1 above)
Comment (AlwinEsch) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: I can change to new with "comma" behind, but to have in one line make for me only sense if only one value, with more and on changes is more hard to see what is changed,

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Game.h (hunk 2 above)
Comment (Rechi) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Game.h]: Seems like GitHub shows the comment in the Conversation tab at at a wrong line. This comment is for line 2140, not for 2065.

File: xbmc/games/addons/GameClientInGameSaves.cpp (hunk 1 above)
Comment (Rechi) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: The decision was mad that multiple values also go into one line.
You even approved that coding style change.

Comment (AlwinEsch) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: My approval was related to the example with:
```cpp
CPVRChannelGroupInternal::CPVRChannelGroupInternal(bool bRadio) : m_iHiddenChannels(0)
{
}
```
and only one value, thats why it was acceptable for me.

Comment (Rechi) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: https://github.com/xbmc/xbmc/compare/35b2594f369a8b79f22471943a4c47d526023dbc...b3eea1f94c0612f39eb49cfb510122ed9485547a was the state of the PR at the time you approved it and it contained already the following.

### 11.6. Constructor Initialzation Lists
For short lines everything on one line is OK, excluding the braces which must be on the following lines.
```cpp
MyClass::CMyClass(bool bBoolArg, int iIntegerArg) : m_bArg(bBoolArg), m_iArg(iIntegerArg)
{
}
```

Comment (AlwinEsch) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: I change it :smile:, but with `For short lines everything on one line is OK` sounds for me more like as **"optional"** :wink:.

**Comment by AlwinEsch** [comment]
Is rebased and checked. From my side Ok to become in.

Only as remind. I or others have yet to find a good way to eliminate the use of `CBinaryAddonCache` and allow the same instance usage as other addons (which eliminates the need of cache for it) and then add in future request.

**Comment by AlwinEsch** [comment]
jenkins build this please

File: xbmc/games/addons/GameClientInGameSaves.cpp (hunk 1 above)
Comment (Rechi) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: This change is still missing.

Comment (AlwinEsch) [reply xbmc/games/addons/GameClientInGameSaves.cpp]: See my note there https://github.com/xbmc/xbmc/pull/16701#issuecomment-544144978
Also is for me the guideline more by a one line and not like this:
```cpp
CGameClientInGameSaves::CGameClientInGameSaves(CGameClient* addon,
                                               const AddonInstance_Game* dllStruct)
  : m_gameClient(addon), m_dllStruct(dllStruct)
{
}
```

**Pull request merged**